testnet/stacks-node/src/burnchains/bitcoin_regtest_controller.rs

```diff
@@ -787,49 +911,137 @@ impl BitcoinRegtestController {
 
         for commit_to in payload.commit_outs.iter() {
             tx.output
-                .push(commit_to.to_bitcoin_tx_out(value_per_transfer));
+                .push(commit_to.to_bitcoin_tx_out(estimated_fees.amount_per_output()));
         }
 
+        let fee_rate = estimated_fees.fee_rate;
         self.finalize_tx(
             &mut tx,
-            budget_for_outputs,
-            btc_miner_fee + rbf_fee,
-            utxos,
+            estimated_fees.total_spent_in_outputs(),
+            estimated_fees.spent_in_attempts,
+            estimated_fees.min_tx_size(),
+            fee_rate,
+            utxos.clone(),
             signer,
-            attempt,
         )?;
 
-        increment_btc_ops_sent_counter();
+        let serialized_tx = SerializedTx::new(tx.clone());
+
+        let tx_size = serialized_tx.bytes.len() as u64;
+        estimated_fees.register_replacement(tx_size);
+        let mut txid = tx.txid().as_bytes().to_vec();
+        txid.reverse();
+        let ongoing_block_commit = OngoingBlockCommit {
+            payload,
+            utxos,
+            fees: estimated_fees,
+            txid: Txid::from_bytes(&txid[..]).unwrap(),
+        };
 
         info!(
-            "Miner node: submitting leader_block_commit op for {} - {}, waiting for its inclusion in the next Bitcoin block",
-            &payload.block_header_hash,
-            public_key.to_hex()
+            "Miner node: submitting leader_block_commit (txid: {}, rbf: {}, total spent: {}, size: {}, fee_rate: {})",
+            ongoing_block_commit.txid.to_hex(),
+            ongoing_block_commit.fees.is_rbf_enabled,
+            ongoing_block_commit.fees.total_spent(),
+            ongoing_block_commit.fees.final_size,
+            fee_rate,
         );
 
+        self.ongoing_block_commit = Some(ongoing_block_commit);
+
+        increment_btc_ops_sent_counter();
+
         Some(tx)
     }
 
+    fn build_leader_block_commit_tx(
+        &mut self,
+        payload: LeaderBlockCommitOp,
+        signer: &mut BurnchainOpSigner,
+        _attempt: u64,
+    ) -> Option<Transaction> {
+        // Are we currently tracking an operation?
+        if self.ongoing_block_commit.is_none() {
+            // Good to go, let's build the transaction and send it.
+            let res = self.send_block_commit_operation(payload, signer, &vec![], &vec![], None);
+            return res;
+        }
+
+        let ongoing_op = self.ongoing_block_commit.take().unwrap();
+
+        let _ = self.sortdb_mut();
+        let mined_op = self
+            .burnchain_db
+            .as_ref()
+            .expect("BurnchainDB not opened")
+            .get_burnchain_op(&ongoing_op.txid);
+
+        if mined_op.is_some() {
```

Comment:
What happens in the following scenario:

1. `outgoing_op` is mined in Block N
2. `build_leader_block_commit_tx()` is called
3. burnchain_db receives Block N

Basically, if the `outgoing_op` is confirmed and this method is called _before_ the `burnchain_db` knows about the confirmation. It seems like this method would try to RBF that transaction, correct? And then, one of two things will happen: (1) the bitcoind instance communicated with would already know that the prior op was confirmed and there'd be an error, or (2) the bitcoind instance hasn't received the new block and accepts the RBF. In this case, the outgoing transaction would just be "wrong" -- the controller would be tracking a transaction that is going to be dropped by the network (which is possible no matter what with RBF). I think to remedy this, the `outgoing_op` struct may need to track multiple TXIDs?

Reply:
I could be missing a subtlety, but I think that the case you mentioned is covered by the next lines, ensuring "idempotency":
```rust
        if payload == ongoing_op.payload {
            debug!("Abort attempt to re-submit identical LeaderBlockCommit");
            return None;
        }
```